pkg/archive/tar.go

```diff
@@ -133,21 +144,21 @@ func ExtractTar(ctx context.Context, d string, f string) error {
 				return fmt.Errorf("failed to close file: %w", err)
 			}
 		case tar.TypeSymlink:
-			// Skip symlinks for targets that do not exist
-			_, err = os.Readlink(target)
+			// Ensure that symlinks are not relative path traversals
+			// #nosec G305 // L158 handles the check
+			fullLink := filepath.Join(d, header.Linkname)
+			_, err = os.Lstat(fullLink)
 			if os.IsNotExist(err) {
 				continue
 			}
-			// Ensure that symlinks are not relative path traversals
-			// #nosec G305 // L208 handles the check
-			linkReal, err := filepath.EvalSymlinks(filepath.Join(d, header.Linkname))
+			linkReal, err := filepath.EvalSymlinks(fullLink)
 			if err != nil {
 				return fmt.Errorf("failed to evaluate symlink: %w", err)
 			}
 			if !IsValidPath(target, d) {
 				return fmt.Errorf("symlink points outside temporary directory: %s", linkReal)
```

Comment:
So before this proposed change, the earlier os.Readlink() call would fail (because the symlink hadn't been created yet), and so for symlinks, they would never get to this. However, this check here should be changed to `if !IsValidPath(linkReal, d)` to ensure that the destination of the symlink is within the extraction directory. Without it, a symlink that points outside the directory will be created:

```
$ git checkout -q egibs/fix-tar-gz-edge-case
$ git show --no-patch --oneline HEAD
581cbe90 (HEAD, egibs/fix-tar-gz-edge-case) Add .gzip for completeness, move above .zip
$ go build  ./cmd/mal/
$ tar tvf ~/tmp/symlink.tar.gz
drwxrwxr-x user/user 0 2024-12-31 13:14 deleteme2/
lrwxrwxrwx user/user 0 2024-12-31 13:14 deleteme2/ls -> ../../../../../../../usr/bin/ls
$ ./mal --format=simple analyze ~/tmp/symlink.tar.gz
# /home/user/tmp/symlink.tar.gz ∴ /usr/bin/ls: medium
c2/addr/url: low
c2/tool_transfer/arch: low
data/compression/lzma: low
discover/system/hostname: low
evasion/file/prefix: medium
exec/shell/TERM: low
fs/link_read: low 
net/url/embedded: low
process/name_set: medium
```

Reply:
Oh, yea, good catch. It's hard to keep `target`,`linkReal, and `header.Linkname` straight.

Reply:
Given that the checks for symlinks are the same for both debs and tarballs, it seems likely that the validation could probably be abstracted into a helper function, so that there's not duplicated code that needs to be kept consistent.

Reply:
Added in [`c0d9a99` (#733)](https://github.com/chainguard-dev/malcontent/pull/733/commits/c0d9a99b2e8bdbbe5cc0106891dbc1b355663f97).

I simplified the permissions; directories are `0o700` for traversal, everything else is `0o600`.

Additionally, `handleFile` will error if we exceed `maxBytes` (this was not evident prior) and `handleSymlink` will ensure that existing symlinks are cleaned up before we create them.